src/Umbraco.Core/Routing/UrlProvider.cs

```diff
@@ -111,8 +111,10 @@ public string GetUrl(IPublishedContent content, UrlMode mode = UrlMode.Default,
             // this the ONLY place where we deal with default culture - IUrlProvider always receive a culture
             // be nice with tests, assume things can be null, ultimately fall back to invariant
             // (but only for variant content of course)
-            if (content.ContentType.VariesByCulture())
+            // We need to check all ancestors because urls are variant even for invariant content, if an ancestor is variant.
+            if (content.AncestorsOrSelf().Any(x=>x.ContentType.VariesByCulture()))
             {
+
                 if (culture == null)
                     culture = _variationContextAccessor?.VariationContext?.Culture ?? "";
             }
```

Comment:
This avoids calling `AncestorsOrSelf()` if we already have a culture set:
```suggestion
            if (culture == null && content.AncestorsOrSelf().Any(x => x.ContentType.VariesByCulture()))
            {
                    culture = _variationContextAccessor?.VariationContext?.Culture ?? "";
            }
```